cpp/react/renderer/components/RNLiveMarkdownSpec/MarkdownTextInputDecoratorShadowNode.mm

```diff
@@ -0,0 +1,256 @@
+#include "MarkdownTextInputDecoratorShadowNode.h"
+
+#include <react/renderer/components/iostextinput/TextInputState.h>
+
+#include <React/RCTUtils.h>
+#include <react/renderer/components/view/conversions.h>
+#include <react/renderer/core/ComponentDescriptor.h>
+#include <react/renderer/textlayoutmanager/RCTAttributedTextUtils.h>
+#include <yoga/Yoga.h>
+
+#include "RCTMarkdownStyle.h"
+#include "RCTMarkdownUtils.h"
+
+using namespace expensify::livemarkdown;
+
+namespace facebook {
+namespace react {
+
+extern const char MarkdownTextInputDecoratorViewComponentName[] =
+    "MarkdownTextInputDecoratorView";
+
+void MarkdownTextInputDecoratorShadowNode::initialize() {
+  ShadowNode::traits_.unset(ShadowNodeTraits::ForceFlattenView);
+}
+
+void MarkdownTextInputDecoratorShadowNode::adoptChildren() {
+  const auto &children = getChildren();
+  if (children.size() == 0) {
+    return;
+  }
+  react_native_assert(
+      children.size() != 1 &&
+      "MarkdownTextInputDecoratorView received more than one child");
+
+  if (const auto child = std::dynamic_pointer_cast<const TextInputShadowNode>(
+          children.at(0))) {
+    // don't mind this :)
+    const auto &nodeWithAccessibleYogaNode =
+        reinterpret_cast<const MarkdownTextInputDecoratorShadowNode *>(&*child);
+
+    // decorator node cannot have a measure function since it's not a leaf node
+    // but we can redirect measuring of the child input to call measureContent
+    // on the decorator
+    YGNodeSetMeasureFunc(&nodeWithAccessibleYogaNode->yogaNode_,
+                         yogaNodeMeasureCallbackConnector);
+  }
+}
+
+void MarkdownTextInputDecoratorShadowNode::appendChild(
+    const ShadowNode::Shared &child) {
+  YogaLayoutableShadowNode::appendChild(child);
+
+  adoptChildren();
+}
+
+void MarkdownTextInputDecoratorShadowNode::replaceChild(
+    const ShadowNode &oldChild, const ShadowNode::Shared &newChild,
+    size_t suggestedIndex) {
+  YogaLayoutableShadowNode::replaceChild(oldChild, newChild, suggestedIndex);
+
+  adoptChildren();
+};
+
+Size MarkdownTextInputDecoratorShadowNode::measureContent(
+    const LayoutContext &layoutContext,
+    const LayoutConstraints &layoutConstraints) const {
+  const auto &children = getChildren();
+  react_native_assert(
+      children.size() == 1 &&
+      "MarkdownTextInputDecoratorView received wrong number of children");
+
+  const auto child =
+      std::static_pointer_cast<const TextInputShadowNode>(children.at(0));
+
+  child->ensureUnsealed();
+
+  // apply markdown formatting before measuring the child
+  const auto &mutableChild =
+      std::const_pointer_cast<TextInputShadowNode>(child);
```

Comment:
The logic here (not specifically these lines, in general) assumes that if the decorator node is mutable, the child text input is also mutable. I'm not actually sure whether this assumption always holds true.

It may make sense to check the children in the constructor and clone them if `fragment.children` is null and the children list is not empty to ensure the child is mutable.